docs/adr/adr-004-state-interaction.md

```diff
@@ -0,0 +1,189 @@
+
+Upon node initialisation, the node will receive the endpoint of a trusted celestia-core node. The constructor for 
+`CoreAccess` will construct a new instance of `CoreAccess` with the `lens.ChainClient` pointing at the given 
+celestia-core endpoint.
+
+```go
+type CoreAccess struct {
+    cc *lens.ChainClient
+}
+
+func NewCoreAccess(cc *lens.ChainClient) (*CoreAccess, error) {
+	return &CoreAccess{
+        cc: cc,		
+    }, nil   
+}
+
+// where Account wraps account information and Balance 
+func (ca *CoreAccess) AccountBalance(ctx context.Context, acct Account) (Balance, error) {
+	coins, err := da.cc.QueryBalanceWithAddress(acct.address)
+	if err != nil {
+		return err
+    }   
+    // returning first index as we only care for account's balance on celestia chain	
+	return coins[0], nil 
+}
+
+// where Tx implements sdk.Msg
+func (ca *CoreAccess) SubmitTx(ctx context.Context, tx Tx) (TxResponse, error) {
+	return da.cc.SendMsg(ctx, tx)
+}
+```
+
+### `P2PAccess`
+
+An idea for how `P2PAccess` could be implemented can be found below.
+
+```go
+type P2PAccess struct {
+	host libhost.Host
+    // P2PAccess will still use lens.ChainClient's methods
+	// under the hood, but will use libp2p transport instead
+	// of RPC communication with the Core endpoint.
+   cc  *lens.ChainClient 
+}
+
+func NewP2PAccess(host libhost.Host, cc *lens.ChainClient) *P2PAccess {
+    return &P2PAccess{
+		host: host,
+		cc: cc,
+    }   	
+}
+
+func (pa *P2PAccess) AccountBalance(ctx context.Context, acct Account) (*Balance, error) {
+    // passing in lens.ChainClient here to help construct the query
+	query := acctBalanceQuery(pa.cc, acct) 
+	
+	// open connection with peer who serves State
+    stream, err := pa.ConnectToStateProvider()   
+	if err != nil {
+		return nil, err
+    }
+    
+	// write query to stream
+    err := stream.Write(query)
+    if err != nil {
+		return nil, err
+    }
+
+	// read resp from stream
+    resp, err := stream.Read(buf)
+    if err != nil {
+        return nil, err		    
+    }	
+
+	// unmarshal balance from response
+	var bal Balance
+    err = bal.Unmarshal(buf)	
+	if err != nil {
+		return nil, err
+    }
+	
+    return bal, nil
+}
+
+// where Tx implements sdk.Msg
+func (pa *P2PAccess) SubmitTx(ctx context.Context, tx Tx) (*TxResponse, error) {
+    // passing in lens.ChainClient here to help construct the query 
+	query := submitTxQuery(pa.cc, tx)
+	
+    // open connection with peer who serves State
+    stream, err := pa.ConnectToStateProvider()
+    if err != nil {
+        return nil, err
+    }
+    
+    // write query to stream
+    err := stream.Write(query)
+    if err != nil {
+		return nil, err
+    }
+
+	// read response from stream
+    resp, err := stream.Read(buf)
+    if err != nil {
+        return nil, err
+    }	
+	
+	// get txResponse from response
+	var tx Tx
+	err = tx.Unmarshal(buf)
+	if err != nil {
+		return nil, err
+}
+	
+    return TxResponse, nil
```

Comment:
What's the point of having streams here if you say that we will do everything with the lens methods above, but with hijacked p2p transport?

Reply:
WE can also manually create the queries.

Reply:
If we go with lens and core.Client then the P2PAccess implementation won't have such logic, but it will use the lens, only core.Client over p2p will have those and not the P2PAcess, so there is discrepancy within the code. You didn't mention the code as pesudo-code and its seems like your goal is use it further, so It makes sense to check if the code is actually workable. Maybe not compile, but at least is doing the right thing.

Reply:
I honestly don't want to detail the p2p implementation here right now bc we don't need to implement it for this iteration anyway and it might change later on. So should I just remove the code snippets for now?

Reply:
removed the code for the most part, pls check again.

Reply:
I am ok to remove it!